### PR TITLE
Segmentation metrics

### PR DESCRIPTION
**Pull request by GFabien**
Add two distance based metrics and allow to compute metrics between channels.

### CODE REVIEW

File: segmentation/utils.py [hunk 1]
```diff
@@ -155,6 +155,26 @@ def channel_metrics(prediction, target, metric):
...
+    channels = list(range(target.shape[1]))
+    res = []
+    for i in channels:
+        for j in channels:
+            if i != j:
+                mask = target[:, j, ...] > cut
+            else:
+                mask = (target[:, j, ...] > 0) * (target[:, j, ...] < 1)
+            res.append(metric(prediction[:, i, ...] * mask, target[:, i, ...] * mask))
+
```
Comment (romainVala) [inline segmentation/utils.py:174]: j'imagine que ca depend des metriques, 
mais en generale multiplier par le mask n'est pas equivalent a ne considerer que les voxels dans le mask
mask_prediction = prediction[mask>cut] 
pour les metrics de SNR par exemple il faut prendre la 2iem options ... non ?

File: segmentation/utils.py [hunk 2]
```diff
@@ -155,6 +155,26 @@ def channel_metrics(prediction, target, metric):
     return torch.tensor(res)
 
 
+def between_channel_metrics(prediction, target, metric, cut=0.5):
```
Comment (romainVala) [inline segmentation/utils.py:158]: pourquoi dans utils, plutot que dans metrics (par exemple metrics/utils_metric.py ou metrics/mask_metrics.py ...)
?

File: segmentation/utils.py [hunk 3]
```diff
@@ -155,6 +155,26 @@ def channel_metrics(prediction, target, metric):
...
+    """
+    Compute a given metric on every channel using masks computed from the other channels.
+    """
```
Comment (romainVala) [inline segmentation/utils.py:161]: j'ai un doute si ta solution est assez générale,
par exemple pour calculer les nombres de faux positif, dans la WM, il faut que la WM soit dans les channels c'est bien ca ?
ca pose pas de problem avec le reste du code (le fait de mettre la WM dans les channels) si on ne fait que la prediction de la GM ?

Comment (GFabien) [reply segmentation/utils.py:161]: Si effectivement, ça pose un problème dans le cas où le modèle n'a été entraîné que sur matière grise / autre.
Tu as une idée pour résoudre le problème sans tout changer à la façon dont on appelle les métriques ? J'ai l'impression qu'il faut fournir à la métrique le masque du tissu qui nous intéresse mais la métrique ne voit que la prédiction et la target. Ou alors il faut donner à la métrique directement la prédiction et la target masquées mais là encore il faut voir comment gérer ça sans tout modifier...

File: segmentation/utils.py (hunk 2 above)
Comment (GFabien) [reply segmentation/utils.py:158]: Parce que j'avais mis les autres fonctions du genre à cet endroit, je suis d'accord, c'est juste que j'avais besoin de la même chose pour les loss (en tout cas de `mean_metric`) donc j'avais mis en commun dans le fichier `utils.py`.

File: segmentation/utils.py (hunk 1 above)
Comment (GFabien) [reply segmentation/utils.py:174]: Pour la plupart des métriques les 0 présents au même endroit dans la prédiction et dans la target ne contribuent pas (il y a les TN mais ça n'a pas de sens de les regarder entre les channels puisque par définition ils sont à 0) et pour certaines métriques la structure est importante (métriques basées sur les distances) qu'on perd avec `mask_prediction = prediction[mask>cut]`.

Pour ce qui est du SNR c'est effectivement plus proche de la 2è option mais c'est quelque chose qui peut se faire dans le calcul de la métrique elle-même d'aller prendre les valeurs non nulles. En revanche je ne suis pas sûr que ça ait du sens d'aller regarder les channels deux à deux pour calculer le SNR.

File: segmentation/utils.py (hunk 2 above)
Comment (romainVala) [reply segmentation/utils.py:158]: ok, mais tu peux faire le changement, c'est logique que les loss, utilise des fonctions de metrics ...

File: segmentation/utils.py (hunk 3 above)
Comment (romainVala) [reply segmentation/utils.py:161]: j'ai pas encore bien en tete la structure du code concernant les metriques, 
mais je dirais de faire un argument mask (optionel, avec default None) dans les metrics ...
quand tu es dans le training tu fais sans le mask, et dans l'eval on peut definir, l'argument mask a ajouter dans metrics (en choisissant le image_key du sample)
...

File: segmentation/utils.py (hunk 1 above)
Comment (romainVala) [reply segmentation/utils.py:174]: effectivement le SNR, c'est un mauvais exemple ...
bon laissons comme ca pour l'instant, et on modifira si besoin 

mais a garde en tete, car  gérer la 2ieme option  dans la metrique n'est pas equivalent : il peut y avoir des zeros dans le masque (qu'il faut prendre en compte) et les zero a l'exterieur du masque qui ne faut pas compter ....

File: segmentation/utils.py (hunk 3 above)
Comment (romainVala) [reply segmentation/utils.py:161]: en gros, au lieu d'avoir le choix des mask dans les channel, il faut un choix explicite avec les clefs du sample, pour pouvoir etres souple
ca fera des definitions plus longue dans les json (car explicite) mais du coup plus flexible
non ?

Comment (GFabien) [reply segmentation/utils.py:161]: Oui (c'est un argument pro tenseurs 3D dans torchio au passage), ce qui est un peu "compliqué" c'est que le sample contient potentiellement un batch et il faut donc récupérer la bonne information mais ce qui est en dehors du scope de la métrique. C'est gérable mais il faut que je fasse des modifications plus profondes. Je te ferai une nouvelle proposition.

File: segmentation/run_model.py [hunk 4]
```diff
@@ -87,11 +89,14 @@ def get_segmentation_data(self, sample):
...
-            targets = to_var(targets[torchio.DATA].float(), self.device)
-        else:
-            targets = None
+        targets = []
+        for key in self.label_key_name:
+            if key in sample:
+                target = sample[key]
+                targets.append(to_var(target[torchio.DATA].float(), self.device))
+            else:
+                return volumes, None
```
Comment (romainVala) [inline segmentation/run_model.py:99]: there should be at least a warning in the log,
en fait pas ici, car sinon on va avoir un message a chaque iteration, mais ce serait bien d'ajouter ce check dans le config, une fois qu'on a construit le dataset (donc on a la list des modalite,) vérifie que la liste des label_key_name correspond, si non, ajout un self.log('WARNING misssing ....')

File: segmentation/run_model.py [hunk 5]
```diff
@@ -87,11 +89,14 @@ def get_segmentation_data(self, sample):
...
+        targets = torch.cat(targets, dim=1)
```
Comment (romainVala) [inline segmentation/run_model.py]: c'est bien comme solution, la liste des label_key_name pour faire la concatenation !
cool

File: segmentation/run_model.py (hunk 4 above)
Comment (GFabien) [reply segmentation/run_model.py:99]: En fait j'ai mis ça car j'utilise le même getter pour l'inférence mais les labels ne sont pas présents à ce moment-là. Je devrais peut-être avoir quelque chose de plus explicite effectivement

Comment (romainVala) [reply segmentation/run_model.py:99]: ca me va comme ca, 
mais je voulais juste qu'on prevoit le cas ou l'utilisateur fait une faute de frappe dans le nom label_key (dans le fichier json), mais c'est annecdotique
